Core/Utilities/Proxy.py

```diff
@@ -183,3 +172,98 @@ def wrapped_fcn( *args, **kwargs ):
       executionLock.release()
 
   return wrapped_fcn
+
+
+class UserProxy(object):
+  """Implement a context manager to execute functions with a user proxy."""
+
+  def __init__(self,
+               proxyUserName=None,
+               proxyUserGroup=None,
+               proxyUserDN=None,
+               proxyWithVOMS=True,
+               proxyFilePath=None,
+               executionLock=False,
+               ):
+    """Construct the context manager for a user proxy.
+
+    :param str proxyUserName: the user name of the proxy to be used
+    :param str proxyUserGroup: the user group of the proxy to be used
+    :param str proxyUserDN: the user DN of the proxy to be used
+    :param bool proxyWithVOMS: optional flag to dress or not the user proxy with VOMS extension ( default True )
+    :param str proxyFilePath: optional file location for the temporary proxy
+    :param bool executionLock: flag to execute with a lock for the time of user proxy application ( default False )
+    """
+    if not ((proxyUserName or proxyUserDN) and proxyUserGroup):
+      self.result = S_OK()
+    else:
+      self.result = _putProxy(userDN=proxyUserDN,
+                              userName=proxyUserName,
+                              userGroup=proxyUserGroup,
+                              vomsFlag=proxyWithVOMS,
+                              executionLockFlag=executionLock,
+                              proxyFilePath=proxyFilePath,
+                              )
+
+  def __enter__(self):
+    return self.result
+
+  def __exit__(self, _type, _value, _traceback):
+    """Reset the proxy and certificate state at the end of the managed block."""
+    if self.result['OK'] and self.result['Value']:
+      originalUserProxy, useServerCertificate, executionLock = self.result['Value']
+      _restoreProxyState(originalUserProxy, useServerCertificate, executionLock)
+
+
+def _putProxy(userDN=None, userName=None, userGroup=None, vomsFlag=None, proxyFilePath=None, executionLockFlag=False):
+  """Download proxy, place in a file and populate X509_USER_PROXY environment variable.
+
+  Parameters like `userProxy` or `executeWithUserProxy`.
+  :returns: Tuple of originalUserProxy, useServerCertificate, executionLock
+  """
+  # Setup user proxy
+  if userDN:
+    userDNs = [userDN]
+  else:
+    result = getDNForUsername(userName)
+    if not result['OK']:
+      return result
+    userDNs = result['Value']  # a same user may have more than one DN
+    vomsAttr = ''
+    if vomsFlag:
+      vomsAttr = getVOMSAttributeForGroup(userGroup)
+
+    result = getProxy(userDNs, userGroup, vomsAttr, proxyFilePath)
+
+    if not result['OK']:
+      return result
+
+    executionLock = LockRing().getLock('_UseUserProxy_', recursive=True) if executionLockFlag else None
+    if executionLockFlag:
+      executionLock.acquire()
+
+    proxyFile = result['Value']
+    originalUserProxy = os.environ.get('X509_USER_PROXY')
+    os.environ['X509_USER_PROXY'] = proxyFile
```

Comment:
Minor, but you can do this switch in just one line:
```
os.environ['X509_USER_PROXY''], originalUserProxy = result['Value'], os.environ.get('X509_USER_PROXY')
```

Reply:
Done